howso/input_validation.amlg

```diff
@@ -249,4 +249,407 @@
+																(= "indices" (current_index 1))
+															)
+															(not (contains_value ["boolean" "number" "string" "list" "assoc" "any"] (current_value) ))
+														)
+														(and
+															;if a string value reference, ensure that reference is in the !customTypes
+															(~ "" (current_value))
+															(= "ref" (current_index))
+															(not (contains_index !customTypes (current_value)))
+														)
+														(and
+															;if an assoc (that's not the value of "indices"), it must have a type or ref
+															(~ (assoc) (current_value))
+															(!= "indices" (current_index))
+															(and
+																(not (contains_index (current_value) "type"))
+																(not (contains_index (current_value) "ref"))
+															)
+														)
+													)
+													(accum (assoc invalid_type_hints (list (concat method_name "/" parameter_name))))
+												)
+											)
+											typing_map
+										)
+									)
+
+									;return the typing map with default/optional added
+									(append
+										typing_map
+										(assoc
+											default default_value
+											optional (= default_value (null))
+										)
+									)
+								)
+							)
+							;first item is the assoc of parameter names to comments/default value
+							(first (get_entity_comments (null) (current_index) (true)))
+						)
+					)
+					(get_entity_comments (null) (null) (true))
+				)
+		))
+
+		;also need to check the CustomTypes schemas
+		(map
+			(lambda
+				(call !CheckTypingMap (assoc
+					typing_map (current_value 1)
+					method_name "CustomTypes"
+					parameter_name (current_index 1)
+				))
+			)
+			!customTypes
+		)
+
+		(if (size invalid_type_hints)
+			;return list of invalid type hints
+			invalid_type_hints
+
+			(assign_to_entities (assoc !parameterValidationMap param_validation_map ))
+		)
+	)
+
+	;Method that uses #!parameterValidationMap to get the appropriate typing specification for each parameter of the called label
+	;and uses this specification to validate the parameter values given as defaults or by the user.
+	;no parameters, but inspects the parameters passed along the label this method was called from.
+	#!ValidateParameters
+	(declare
+		(assoc
+			;this automatically determines the label/method that this method is being called within
+			label_name (first (get_labels (get (opcode_stack) -8)))
+			;the args passed to the method above
+			given_parameters (args 1)
+		)
+
+		(declare (assoc
+			undefined_parameters (indices (remove given_parameters (indices (get !parameterValidationMap label_name))))
+		))
+
+		(if (size undefined_parameters)
+			(conclude (conclude
+				(call !Return (assoc
+					errors
+						(list (concat
+							"The following given parameters are not supported for " label_name ": "
+							(apply "concat" (trunc (weave undefined_parameters ", ")))
```

Comment:
Will want to make this is more robust once the clients are updated. This will eventually need to produce a general `detail` error message and provide an `errors` object containing the structure defined in ticket #20103. Will need to map out what `codes` we want to use for each error type when we get there.

Reply:
yep, sounds good. Will do then